packages/main/src/Token.ts

```diff
@@ -0,0 +1,243 @@
+ *
+ * <h3>ES6 Module Import</h3>
+ *
+ * <code>import "@ui5/webcomponents/dist/Token.js";</code>
+ * @constructor
+ * @author SAP SE
+ * @alias sap.ui.webc.main.Token
+ * @extends sap.ui.webc.base.UI5Element
+ * @tagname ui5-token
+ * @since 1.0.0-rc.9
+ * @implements sap.ui.webc.main.IToken
+ * @public
+ */
+@customElement("ui5-token")
+@languageAware
+
+/**
+ * Fired when the the component is selected by user interaction with mouse or by clicking space.
+ *
+ * @event sap.ui.webc.main.Token#select
+ * @public
+ */
+@event("select")
+
+/**
+ * Fired when the backspace, delete or close icon of the token is pressed
+ *
+ * @event
+ * @param {Boolean} backSpace Indicates whether token is deleted by backspace key.
+ * @param {Boolean} delete Indicates whether token is deleted by delete key.
+ * @private
+ */
+@event("delete", {
+	detail: {
+		"backSpace": { type: Boolean },
+		"delete": { type: Boolean },
+	},
+})
+
+class Token extends UI5Element implements ITabbable {
+	/**
+	 * Defines the text of the token.
+	 *
+	 * @type {string}
+	 * @name sap.ui.webc.main.Token.prototype.text
+	 * @defaultvalue ""
+	 * @public
+	 */
+	@property()
+	text!: string;
+
+	/**
+	 * Defines whether the component is read-only.
+	 * <br><br>
+	 * <b>Note:</b> A read-only component can not be deleted or selected,
+	 * but still provides visual feedback upon user interaction.
+	 *
+	 * @type {boolean}
+	 * @name sap.ui.webc.main.Token.prototype.readonly
+	 * @defaultvalue false
+	 * @public
+	 */
+	@property({ type: Boolean })
+	readonly!: boolean;
+
+	/**
+	 * Set by the tokenizer when a token is in the "more" area (overflowing)
+	 * @type {boolean}
+	 * @defaultvalue false
+	 * @private
+	 */
+	@property({ type: Boolean })
+	overflows!: boolean;
+
+	/**
+	 * Defines whether the component is selected or not.
+	 *
+	 * @type {boolean}
+	 * @name sap.ui.webc.main.Token.prototype.selected
+	 * @defaultvalue false
+	 * @public
+	 */
+	@property({ type: Boolean })
+	selected!: boolean;
+
+	/**
+	 * Defines whether the component is focused or not.
+	 *
+	 * @type {boolean}
+	 * @defaultvalue false
+	 * @private
+	 */
+	@property({ type: Boolean })
+	focused!: boolean;
+
+	/**
+	 * Defines whether the token is being deleted
+	 * This flag is used in the ui5-multi-combobox
+	 *
+	 * @type {boolean}
+	 * @defaultvalue false
+	 * @private
+	 */
+	@property({ type: Boolean })
+	toBeDeleted!: boolean;
+
+	/**
+	 * Defines the tabIndex of the component.
+	 * @type {string}
+	 * @private
+	 */
+	@property({ defaultValue: "-1", noAttribute: true })
+	_tabIndex!: string;
+
+	/**
+	 * Defines the close icon for the token. If nothing is provided to this slot, the default close icon will be used.
+	 * Accepts <code>ui5-icon</code>.
+	 *
+	 * @type {sap.ui.webc.main.IIcon[]}
+	 * @name sap.ui.webc.main.Token.prototype.closeIcon
+	 * @slot closeIcon
+	 * @public
+	 * @since 1.0.0-rc.9
+	 */
+	@slot()
+	closeIcon!: Array<HTMLElement>;
```

Comment:
why marked as an array? It used to be a single Icon

https://github.com/SAP/ui5-webcomponents/blob/main/packages/main/src/Token.js#L96-L104

Reply:
All slots are array by design, so it's correct

Reply:
I understand that but in most of the "single" use slots (such as Icon or Page's header / foorter) we consider that slot should contain a single element. Which means when used as an array, e.g. 2 or 3 icons are passed here layouting might break. Is this acceptable? Isn't that migration to TS kind of a guide to our users how to correctly use the components?

Reply:
Yes, everyone can pass multiple icons, there is no restriction - in HTML you usually can pass random number of elements. It probably won't work, that's why we document it the way it's meant to be used, and if someone abuses it - it's not a case that we support. It's allowed, we don't throw, and if visually it's bad, it's on the consumer's risk. We state what we support in the API.

At the moment, we can't write anything else here if we don't change the framework, because it's physically an array.
Several times this was discussed and we don't change this concept, because the way it is now is closer to the native behaviour of the html elements.

The TS migration is for the component developers, not for the users. TS helps the developers to make less mistakes, because of static checks, and the code more readable and descriptive. For the users, the same documentation and usage is displayed as before.

Reply:
We used to have such a notion for slots at the very beginning of the project - single/multiple, but we abandoned it, because HTML does not "support" this by design. The framework expects that all slots are arrays, so technically we didn't change anything.